packages/core/src/graphql-module.ts

```diff
@@ -197,74 +202,61 @@ export class GraphQLModule<Config = any, Session = any, Context = any> {
-      const selfDirectiveResolvers = this.selfDirectiveResolvers;
-      if (selfDirectiveResolvers) {
-        directiveResolversSet.add(selfDirectiveResolvers);
-      }
-      this._cache.directiveResolvers = deepmerge.all([...directiveResolversSet]) as IDirectiveResolvers;
-    }
-    return this._cache.directiveResolvers;
-  }
-
-  buildSchemaWithMergeSchemas() {
-    const modulesMap = this.modulesMap;
+  private buildSchemaWithMergeSchemas() {
     const schemaSet = new Set<GraphQLSchema>();
-    this.iterateOverImports(module => {
-      module.fixCacheForModulesMap(modulesMap, 'schema');
+    const selfImports = this.selfImports;
+    for (const module of selfImports) {
+      if (typeof module === 'undefined') {
+        throw new DependencyModuleUndefinedError(this.name);
+      }
       const moduleSchema = module.schema;
       if (moduleSchema) {
         schemaSet.add(moduleSchema);
       }
-    });
+    }
     const selfTypeDefs = this.selfTypeDefs;
-    if (selfTypeDefs) {
-      schemaSet.add(buildASTSchema(selfTypeDefs, {
-        assumeValid: true,
-        assumeValidSDL: true,
-      }));
+    let resolvers = {};
+    const schemaDirectives = this.schemaDirectives;
+    try {
+      if (selfTypeDefs) {
+        const localSchema = buildASTSchema(selfTypeDefs, {
+          assumeValid: true,
+          assumeValidSDL: true,
+        });
+        schemaSet.add(localSchema);
+      }
+      const selfResolversComposition = this.selfResolversComposition;
+      if (Object.keys(selfResolversComposition).length) {
+        resolvers = this.resolvers;
+      } else {
+        resolvers = this.addSessionInjectorToSelfResolversContext();
+      }
+      const selfExtraSchemas = this.selfExtraSchemas;
+      const schemas = [...selfExtraSchemas, ...schemaSet];
+      if (schemas.length) {
+        this._cache.schema = mergeSchemas({
+          schemas,
+          resolvers,
+          schemaDirectives,
+        });
+      } else {
+        this._cache.schema = null;
+      }
+    } catch (e) {
+      if (e.message.includes(`Type "`) && e.message.includes(`" not found in document.`)) {
+        const typeDef = e.message.replace('Type "', '').replace('" not found in document.', '');
+        throw new TypeDefNotFoundError(typeDef, this.name);
+      } else {
+        throw new SchemaNotValidError(this.name, e.message);
+      }
     }
-    const selfExtraSchemas = this.selfExtraSchemas;
-    const selfResolvers = this.addSessionInjectorToSelfResolversContext();
-    const selfResolversComposition = this.addSessionInjectorToSelfResolversCompositionContext();
-    const selfComposedResolvers = composeResolvers(
-      selfResolvers,
-      selfResolversComposition,
-    );
-    const selfSchemaDirectives = this.selfSchemaDirectives;
-    this._cache.schema = mergeSchemas({
-      schemas: [ ...selfExtraSchemas, ...schemaSet],
-      resolvers: selfComposedResolvers,
-      schemaDirectives: selfSchemaDirectives,
-    });
   }
-
+/*
```

Comment:
?

Reply:
Anyway I reverted it back :) We need to implement `mergeSchemas` logic later; because the one from `graphql-tools` doesn't add directives and scalars.